File: py/bemane.py
#!/usr/bin/env python3
"""A short script to keep a dir of BEMA episodes up-to-date"""

from pathlib import Path

import requests
from bs4 import BeautifulSoup
from bs4.element import Tag


def weirdo_title(title: str) -> str:
    """A couple of the titles (as of 2022-02-09) are non-numbered. We give them
    b-numbers, copying the website's URL schema and putting them in their
    correct place in the sequence."""
    if title == '"Jesus Shema"':
        return "021b: " + title
    if title == "Verastikh (Hosea)":
        return "049b: " + title
    return title


def pad_title(title: str) -> str:
    """Zero-pad the number in the titles so that
    Roku Media Player sorts correctly"""
    parts = title.split(":")
    number, title_name = parts[0], ":".join(parts[1:])
    return number.zfill(3) + ":" + title_name


def handle_title(title: str) -> Path:
    "There's a few different quirks in the titles that need handled"
    if "/" in title:
        title = title.replace("/", "_")

    if ":" in title:
        new_title = pad_title(title)
    else:
        new_title = weirdo_title(title)

    return Path(new_title + ".mp3")


def downloader(item: Tag) -> str:
    """Download a given file"""
    filename = handle_title(item.title.text)
    url = item.enclosure["url"]
    if not filename.exists():
        with requests.get(url, stream=True) as request:
            request.raise_for_status()
            with open(filename, "wb") as filehandle:
                for chunk in request.iter_content(chunk_size=8192):
                    filehandle.write(chunk)

    return filename


def main() -> None:
    """Entrypoint"""
    rss = requests.get("https://www.bemadiscipleship.com/rss")
    rss.raise_for_status()
    soup = BeautifulSoup(rss.text, "html.parser")
    channel = soup.rss.channel
    for item in channel.find_all("item"):
        downloader(item)


if __name__ == "__main__":
    main()

File: julia/src/bemane.jl
import HTTP, EzXML

function main()
    response = HTTP.get("https://www.bemadiscipleship.com/rss")
    parsed = parsexml(String(response.body))
    elements = root(parsed)
    items = findall("rss/channel/item", elements)
    for item in items
        downloader(item)
    end
end

function pad_title(title::AbstractString)
    parts = split(title, ':')
    number = parse(UInt16, shift!(parts))
    string(lpad(number, 3, '0'), ':', join(parts, ""))
end

function weirdo_title(title::AbstractString)
    if title == "“Jesus Shema”"
        return string("021b: " title)
    elseif title == "Verastikh (Hosea)"
        return string("049b: " title)
    end
end

function handle_title(title::AbstractString)
    deslashed = replace(title, "/" => "_")
    if ':' in title
        pad_title(title)
    else
        weirdo_title(title)
    end
end

function downloader(item::EzXML.Node)
    title = nodecontent(findfirst("title"))
    filename = handle_title(title)
    enclosure_attrs = attributes(findfirst("enclosure", item))
    url = nodecontent(enclosure_attrs[1])
    if !file_exists(filename)
        fetch(filename)
    end
end

main()
